File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/pythonworker/promisehandlers/ModifyPythonLogicHandler.scala
package edu.uci.ics.amber.engine.architecture.pythonworker.promisehandlers

import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand

object ModifyPythonLogicHandler {
  final case class ModifyPythonLogic(code: String, isSource: Boolean) extends ControlCommand[Unit]
}

File: core/amber/src/main/python/core/architecture/handlers/modify_python_logic_handler.py
from proto.edu.uci.ics.amber.engine.architecture.worker import ModifyPythonLogicV2
from .handler_base import Handler
from ..managers.context import Context
from ...udf import UDFOperator
from ...udf.udf_util import load_udf


class ModifyPythonLogicHandler(Handler):
    cmd = ModifyPythonLogicV2

    def __call__(self, context: Context, command: ModifyPythonLogicV2, *args, **kwargs):
        original_internal_state = context.dp._udf_operator.__dict__
        udf_operator: type(UDFOperator) = load_udf(command.code)
        context.dp._udf_operator = udf_operator()
        # overwrite the internal state
        context.dp._udf_operator.__dict__ = original_internal_state
        return None
